shell/platform/windows/direct_manipulation.cc

```diff
@@ -0,0 +1,225 @@
+    FML_LOG(ERROR) << #operation << " failed"; \
+  }
+
+namespace flutter {
+
+STDMETHODIMP DirectManipulationEventHandler::QueryInterface(REFIID iid,
+                                                            void** ppv) {
+  if ((iid == IID_IUnknown) ||
+      (iid == IID_IDirectManipulationViewportEventHandler)) {
+    *ppv = static_cast<IDirectManipulationViewportEventHandler*>(this);
+    AddRef();
+    return S_OK;
+  } else if (iid == IID_IDirectManipulationInteractionEventHandler) {
+    *ppv = static_cast<IDirectManipulationInteractionEventHandler*>(this);
+    AddRef();
+    return S_OK;
+  }
+  return E_NOINTERFACE;
+}
+
+HRESULT DirectManipulationEventHandler::OnViewportStatusChanged(
+    IDirectManipulationViewport* viewport,
+    DIRECTMANIPULATION_STATUS current,
+    DIRECTMANIPULATION_STATUS previous) {
+  if (current == DIRECTMANIPULATION_RUNNING) {
+    if (!resetting_) {
+      // Not a false event
+      if (owner_->binding_handler_delegate) {
+        owner_->binding_handler_delegate->OnPointerPanZoomStart(
+            (int32_t) reinterpret_cast<int64_t>(this));
+      }
+    }
+  } else if (previous == DIRECTMANIPULATION_RUNNING) {
+    if (resetting_) {
+      // The resetting transition has concluded
+      resetting_ = false;
+    } else {
+      if (owner_->binding_handler_delegate) {
+        owner_->binding_handler_delegate->OnPointerPanZoomEnd(
+            (int32_t) reinterpret_cast<int64_t>(this));
+      }
+      // Need to reset the content transform to its original position
+      // so that we are ready for the next gesture
+      // Use resetting_ flag to prevent sending reset also to the framework
+      resetting_ = true;
+      RECT rect;
+      HRESULT hr = viewport->GetViewportRect(&rect);
+      if (FAILED(hr)) {
+        FML_LOG(ERROR) << "Failed to get the current viewport rect";
+        return E_FAIL;
+      }
+      hr = viewport->ZoomToRect(rect.left, rect.top, rect.right, rect.bottom,
+                                false);
+      if (FAILED(hr)) {
+        FML_LOG(ERROR) << "Failed to reset the gesture using ZoomToRect";
+        return E_FAIL;
+      }
+    }
+  }
+  return S_OK;
+}
+
+HRESULT DirectManipulationEventHandler::OnViewportUpdated(
+    IDirectManipulationViewport* viewport) {
+  return S_OK;
+}
+
+HRESULT DirectManipulationEventHandler::OnContentUpdated(
+    IDirectManipulationViewport* viewport,
+    IDirectManipulationContent* content) {
+  float transform[6];
+  HRESULT hr = content->GetContentTransform(transform, ARRAYSIZE(transform));
+  if (FAILED(hr)) {
+    FML_LOG(ERROR) << "GetContentTransform failed";
+    return S_OK;
+  }
+  if (!resetting_) {
+    // DirectManipulation provides updates with very high precision. If the user
+    // holds their fingers steady on a trackpad, DirectManipulation sends
+    // jittery updates. This calculation will reduce the precision of the scale
+    // value of the event to avoid jitter
```

Comment:
@moffatman What kind of machine have you tested this on?  I'm wondering the feel will vary by device.

Reply:
CC @gspencergoog, who I believe has a Surface set up?